src/contract/event.rs

```diff
@@ -0,0 +1,298 @@
+//! Module implements type-safe event streams from an ABI event definition with
+//! detokenization of the data included in the log.
+
+use crate::abicompat::AbiCompat;
+use crate::errors::{EventError, ExecutionError};
+use crate::log::LogStream;
+use ethcontract_common::abi::{Event as AbiEvent, RawLog};
+use futures::stream::Stream;
+use pin_project::{pin_project, project};
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use web3::api::Web3;
+use web3::contract::tokens::Detokenize;
+use web3::types::{Address, BlockNumber, FilterBuilder};
+use web3::Transport;
+
+/// A type representing a contract event.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Event<T> {
+    /// A new event was received.
+    Added(T),
+    /// A previously revent was removed as the result of a re-org.
+    Removed(T),
+}
+
+impl<T> Event<T> {
+    /// Get the underlying event data regardless of whether the event was added
+    /// or removed.
+    pub fn into_data(self) -> T {
+        match self {
+            Event::Added(value) => value,
+            Event::Removed(value) => value,
+        }
+    }
+
+    /// Get a reference the underlying event data regardless of whether the
+    /// event was added or removed.
+    pub fn as_data(&self) -> &T {
+        match self {
+            Event::Added(value) => value,
+            Event::Removed(value) => value,
+        }
+    }
+
+    /// Get a mutable reference the underlying event data regardless of whether
+    /// the event was added or removed.
+    pub fn as_data_mut(&mut self) -> &mut T {
+        match self {
+            Event::Added(value) => value,
+            Event::Removed(value) => value,
+        }
+    }
+
+    /// Get the underlying event data if the event was added.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if the instance is a removed event.
+    pub fn added(self) -> T {
```

Comment:
Curious to hear your thoughts why panicking is better than making the signature return Option<T>?

Reply:
This is kind of an "unwrap" utility method. This is going to change really soon so I'll leave it as it for now.